Testing the Lambert stuff by using flags

MaqOwais opened the pull request:
Tried to make a better version according to the latest updates as well as removing some errors accordingly

#14972,#18759,#20537,#20545, #20613, #20705

* solveset
  * improved lambert for S.Complexes and S.Reals
  * improvement of  _is_lambert
  * extended to trigonometric and hyperbolic functions
  * test cases

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

MaqOwais committed:
Testing the Lambert stuff by using flags

MaqOwais committed:
quality

MaqOwais committed:
Update bivariate.py

MaqOwais committed:
final update

czgdp1807 commented:
Can this PR in someway close https://github.com/sympy/sympy/pull/14972

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -1105,7 +1122,86 @@ def _solveset(f, symbol, domain, _check=False):
+                                        return result
+                                    if not indls[0].has(Symbol):
+                                        if not indls[0].is_real and domain.is_subset(S.Reals):
+                                            return S.EmptySet
```
When your code ends being indented as much as this it's a clear sign that you need a better structure. Probably what is needed is to define some meaningful functions that can split this up.

MaqOwais commented:
@oscarbenjamin 
Is passing a flag(dmn )  to maintain consistency in domain, a better solution? 
Or is there any other suggestion ?

oscarbenjamin commented:
We should not change the signature of the solve function.

Possibly what is really needed here is to redo the Lambert solving code in the style of the solveset solvers: there should be new code for solveset rather than solveset attempting to use the code from solve. Really the Lambert code works I think by identifying certain forms like `x*exp(x)-a` and also by identifying related forms that can be reduced to that case using a substitution. In `solveset` there is the `substitution` function which I think is designed to handle this sort of thing so it should be straight-forward to build this logic up in the style of solveset.

MaqOwais commented:
> We should not change the signature of the solve function
On adding a random flag (domain -->dnm)  does the signature of solve changes !?
I think the functionality remains same and I think it's like an internal api...

MaqOwais committed:
Do not merge..

I thought of taking 3 stages to complete this PR.
1. This one.
2. Rectify the complexes failed cases in solve with using flags ( it may take upto a week since I'm busy these days in internals)
3. Improving the solvify(since it returns the solution in accordance
with the `solve` output API) and replace solve (in lambert function and other places) in bivariate.py and solveset.py .Hence finally getting rid of solve in solveset and also using flags.( It may take upto a month)

MaqOwais committed:
improved form stage(2)

MaqOwais committed:
updated

MaqOwais committed:
Update solvers.rst

MaqOwais committed:
Update test_wester.py

MaqOwais committed:
Update test_wester.py

MaqOwais committed:
final commit for stage (2)

MaqOwais committed:
Merge branch 'testing-other-stuff' of https://github.com/MaqOwais/sympy into testing-other-stuff

MaqOwais commented:
@oscarbenjamin please give more suggestions to improve this further!

MaqOwais committed:
refactoring solveset and bivariate.py

oscarbenjamin commented on doc/src/tutorial/solvers.rst:
```diff
@@ -179,7 +179,7 @@ multiplicity 1 and ``3`` is a root of multiplicity 2.
 
    >>> solve(x*exp(x) - 1, x )
-   [W(1)]
+   [W(1, integer)]
```
The output from `solve` should not have this "integer" symbol in it.

MaqOwais replied:
instead of integer can I replace it with `n` ?

oscarbenjamin commented:
While this does look like an improvement I think that you need to take account of the basic design of both `solve` and `solveset`. It is difficult for `solve` and `solveset` to share code because they work in different ways and have different expectations:

1. The solve function can depend on the assumptions of the symbols and always returns a finite set of solutions. It has tricks for representing the solutions of equations like `sin(x)=0` but I don't think we can apply those to LambertW because the solutions are not really periodic.
2. The `solveset` function must never depend on the assumptions of the "unknown" symbol as that is given by the domain argument to `solveset`. The `solveset` function must always return a Set that fully represents the set of all possible solutions to the equation.

Because of these differences I think that the best approach here is to leave the current `bivariate` code as it is (or perhaps make small improvements for correctness) but only use it as part of `solve`. Instead new functions should be made for handling these kinds of equations as part of `solveset`.

MaqOwais commented:
> The solveset function must never depend on the assumptions of the "unknown" symbol as that is given by the domain argument to solveset. The solveset function must always return a Set that fully represents the set of all possible solutions to the equation.

I think everything is clear when it comes to `solveset`.(As it's returning right and complete solutions).
And whereas you've said `solve` should not return a solution consisting of a dummy('n') or any symbol ( in this case
 `integer = Symbol('integer',integer=True)` ) but just it should returns only numericals. (I think similar to solveset_reals but imaginary part should also be considered.)

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +183,83 @@ def _lambert(eq, x):
+        if not domain.is_subset(S.Reals):
+            rhs = -c/b + (a/d)*LambertW(arg,integer)
+
+        if not len(list(eq.atoms(Symbol,Dummy))) >1:
+            if domain.is_subset(S.Reals):
+                if -1/E <= re(arg) < 0 and LambertW(arg).is_real:
+                    for k in lambert_real_branches:
+                        rhs = -c/b + (a/d)*LambertW(arg, k)
```
This ode and the lines below are messy and repetitive which makes it hard to understand what the actual logic here is doing. That should be simplified.

MaqOwais commented:
> I think everything is clear when it comes to solveset.(As it's returning right and complete solutions).
What do you think about this?

oscarbenjamin commented:
> > I think everything is clear when it comes to solveset.(As it's returning right and complete solutions).
> 
> What do you think about this?

The code is too messy for me to understand exactly what it is doing so I'm not sure if it is correct.

At least this output is not correct:
```python
In [15]: solveset(x**2*exp(x**2) - x, x, Reals)
Out[15]: 
⎧ -W(2) ⎫
⎪ ──────⎪
⎨   2   ⎬
⎪ℯ      ⎪
⎩       ⎭

In [16]: _.n()
Out[16]: {0.652918640419205}
```
Clearly `x=0` is a solution as well.

MaqOwais commented:
> Clearly x=0 is a solution as well.

Yes, there are some solutions that are not returning the complete solution ...
But as compared to previous commits this has better results (like previously the `solveset` was not returning complete solutions with Imageset).

What is your opinion about using `dmn` flag?

MaqOwais commented:
> And whereas you've said solve should not return a solution consisting of a dummy('n') or any symbol ( in this case
> integer = Symbol('integer',integer=True) ) but just it should returns only numericals. (I think similar to solveset_reals but imaginary part should also be considered.)

Should I consider this point while reviewing again about `solve`? Or any other points to be considered here?

MaqOwais committed:
removed integer from solvers and improved lambert both in solveset(by removing solve from here) and solvers(adding lambert solutions to it where incomplete solutions where present)

MaqOwais committed:
removed integer from solvers and improved lambert both in solveset

MaqOwais committed:
Update test_solveset.py

MaqOwais committed:
simplified lambert stuff

MaqOwais committed:
code quality (refactored and simplified code in bivariate and solveset )

MaqOwais commented:
@oscarbenjamin please do share your opinion on this.

oscarbenjamin commented:
This makes messy special case changes to the main solve and solveset functions that should not be there. It shouldn't be necessary to change so much of the existing functions to make this work.

MaqOwais commented:
> It shouldn't be necessary to change so much of the existing functions to make this work.

Changes here are made to make `solveset` independent of `solve`. The previous changes were actually made by me since I was not aware that solve should not work with `solveset`. Therefore I tried to make the situation better by taking to the previous stage with modifications (i.e. `solveset` returning imageset for complex domain).

oscarbenjamin commented on sympy/solvers/solvers.py:
```diff
@@ -1093,6 +1092,13 @@ def _has_piecewise(e):
     ###########################################################################
     if bare_f:
         solution = _solve(f[0], *symbols, **flags)
+        if solution is not None:
+            for solnArgs in solution:
+                if not isinstance(solnArgs,(dict,tuple,list,set)):
+                    if solnArgs.has(LambertW) and solnArgs.has(Lambda):
+                        return solution
+                        break
```
It shouldn't be necessary to add this kind of special case code in the main `solve` function.

MaqOwais replied:
I think it could be removed.

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -1123,6 +1152,11 @@ def _solveset(f, symbol, domain, _check=False):
             result = FiniteSet(*[s for s in result
                       if isinstance(s, RootOf)
                       or domain_check(fx, symbol, s)])
+        elif isinstance(result,list):
```
I think result here should be a Set. If it can be a list here then something has gone wrong somewhere

MaqOwais replied:
Actually, this code was supplemented along with `solve` function(since `solve` return in form of a list...
yes, this should also be removed.

MaqOwais replied:
This is due to the helper function from `_solve_as_bivariate ` (Line 1809)

MaqOwais commented on sympy/solvers/solveset.py:
```diff
@@ -1748,6 +1782,225 @@ def _is_logarithmic(f, symbol):
+        if g != lhs:
+            usol = _solveset(p, u, domain)
+            if not isinstance(usol, ConditionSet):
+                result = [_solveset(g - us, symbol, domain) for us in usol]
```
here

MaqOwais committed:
refactored accordingly

MaqOwais commented:
@oscarbenjamin please let me know if any other fallacies are present.
I think most of the things are covered up to this point.

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -151,7 +154,11 @@ def _lambert(eq, x):
+        dmn = False
+    else:
+        dmn = True
+    xusolns = solve(X1 - u, x,dmn=dmn)
```
There shouldn't be a `dmn` flag to solve.

MaqOwais replied:
On adding a random flag (domain -->dnm<for domain consistency>) does the signature of solve change!?
I think the functionality remains the same and I think it's like an internal API...

Also, I think the alternate should be present instead of `solve`{imo that could be `solvify` function } to maintain consistent domain.

MaqOwais replied:
As of now I not able to find any other means for domain consistency.
I've tried 
1. global variable
2. passing domain to `solve`, `_tsolve`, `_solve`
3. presently, this flag as `dmn`

oscarbenjamin commented on sympy/solvers/solvers.py:
```diff
@@ -1705,7 +1705,15 @@ def _expand(p):
         try:
             soln = _tsolve(f_num, symbol, **flags)
             if soln is not None:
-                result = soln
+                for solnArgs in soln:
+                    if solnArgs.has(LambertW) and solnArgs.has(Lambda):
+                        return soln
+                        break
+                    else:
+                        result = soln
+            if result is False and soln == []:
+                return soln
```
I don't know what this is doing but I don't think it should be here. Why isn't this just being handled in `_tsolve`?

MaqOwais replied:
I think it could be removed. This was the case for lambert on lambert case. (But that should be dealt with another way.)

oscarbenjamin commented:
You need to resolve the merge conflicts in order for the tests to run

oscarbenjamin commented:
I find this PR difficult because the new code added is messy. Some of the existing code (e.g. solve) is already messy but I don't want to make that worse. It's hard enough to understand what the original code does and I don't understand what the new code does either.

MaqOwais committed:
Merge branch 'master' into testing-other-stuff

MaqOwais committed:
more errors

MaqOwais committed:
Merge branch 'testing-other-stuff' of https://github.com/MaqOwais/sympy into testing-other-stuff

MaqOwais committed:
removed solve totally from solveset and lambert also flag 'dmn' is removed successfully

MaqOwais committed:
updated

MaqOwais committed:
updated (flag 'dmn' is removed)

MaqOwais committed:
will work after improvement of solvify little more

MaqOwais commented:
I wonder why this test has failed. Eventhough, locally everything is fine!

oscarbenjamin commented:
That's #20933

MaqOwais commented:
Please do give your opinion on the latest commit.

oscarbenjamin commented:
Can you explain what all of the changes are?

It's hard to understand why each particular file has been modified when I look in the diff:
https://github.com/sympy/sympy/pull/20824/files

MaqOwais commented:
Changes for 

1. sympy/solvers/bivariate.py
i. Replaced solve with `solvify`. (L157 --> `xusolns` = `solvify`(X1 - u, x,domain) )
ii. added domains to all functions in which are mostly under solveset ( where ever the function is present)
iii. improved output for different cases ( L203-254)

2. sympy/solvers/inequalities.py
i. just rearranged the lines of code. (since at present `solvify` is returning solutions for intersection set whereas previously it was returning [] hence was raising the `ValueError`)

3. sympy/solvers/solvers.py
i. replacing integer with 0 in order to get solutions ( L2702 - 2723)

4. sympy/solvers/solveset.py
i. Most of the lines of code are from issue #14972
ii. Changes in `solvify` are made so that it may give the solution for the intersection set (which could be used with 1. i. replacement of `solve` with `solvify`)
iii. changes in `def _solve_using_known_values(result, solver):` are made to make the test cases in `test_nonlinsolve_using_substitution1` get synchronized with the existing lambert ( it should also return lambert solutions as per problem)

5. Whereas the test cases are modified according to the status of the changes above (1 - 4 points).

MaqOwais commented:
> Changes for
> 
> 1. `sympy/solvers/bivariate.py`
>    i. Replaced solve with `solvify`. (L157 --> `xusolns = solvify(X1 - u, x,domain)` )
>    ii. added domains to all functions in which are mostly under `solveset` ( where ever the function is present)
>    iii. improved output for different cases ( L203-254)
> 
> 2.` sympy/solvers/inequalities.py`
> i. just rearranged the lines of code. (since at present `solvify` is returning solutions for intersection set whereas previously it was returning [] hence was raising the `ValueError`)
> 
> 1. `sympy/solvers/solvers.py`
>    i. replacing `integer` with 0 in order to get solutions as per `solve`( L2702 - 2723)
> 2. `sympy/solvers/solveset.py`
>    i. Most of the lines of code are from issue #14972
>    ii. Changes in `solvify` are made so that it may give the solution for the intersection set (which could be used with 1. i. replacement of `solve` with `solvify`)
>    iii. changes in `def _solve_using_known_values(result, solver):` are made to make the test cases in `test_nonlinsolve_using_substitution1` get synchronized with the existing lambert ( it should also return lambert solutions as per problem)
> 3. Whereas the test cases are modified according to the status of the changes above (1 - 4 points).


I hope this gives a better picture.

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +179,85 @@ def _lambert(eq, x):
+    integer = Symbol('integer',integer=True)
+    for arg in args:
+        rhs1 = -c/b + (a/d)*LambertW(arg)
+        LambertReal = LambertW(arg).is_real
```
A capitalised name like `LambertReal` would normally be used for a class. Since this is a boolean variable it would be better to call this `lambert_real`.

oscarbenjamin replied:
Same goes for e.g. `domainReal`

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +179,85 @@ def _lambert(eq, x):
+        return S.EmptySet
+    # for simplicity defined two functions
+    # Lambert_Real_Braches , lambert_helper
+    def Lambert_Real_Braches():
```
It would be better to use all lower case in function names

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +179,85 @@ def _lambert(eq, x):
+        return sol
+
+    domainReal = domain.is_subset(S.Reals)
+    integer = Symbol('integer',integer=True)
```
This should be a Dummy because this would potentially conflict with a symbol in the input equation

MaqOwais replied:
a symbol can be replaced but Dummy can't be replaced...So it's difficult/not possible to work dummy in such cases...

MaqOwais replied:
or we should assign such a symbol or any such variable that will not conflict..

oscarbenjamin replied:
The function could return the dummy symbol along with the solution.

oscarbenjamin replied:
In fact the best might be to return an `ImageSet`.

MaqOwais replied:
> The function could return the dummy symbol along with the solution.
> In fact the best might be to return an ImageSet.

I agree and that's what is happening right now...
but in between the functions, there is a need for a replaceable variable as the dummy cant be replaced.

oscarbenjamin replied:
A Dummy can be replaced if you use the exact same Dummy. That's why we need to return the Dummy that is used. An ImageSet is a good way of doing this because it bundles the Dummy up with the expression and the domain altogether.

MaqOwais replied:
> A Dummy can be replaced if you use the exact same Dummy. 

I think it's not possible I've tried the index will be changed...
once a dummy is assigned e.g. n it turns into _n then how can we replace it ?

> An ImageSet is a good way of doing this because it bundles the Dummy up with the expression and the domain altogether.
to do this we have to consider lambda expression and after that, we can't replace it with 0 (as it is used in solve<to replace it with 0 to get numerical value >)

MaqOwais replied:
> but in between the functions,

since the dummy variable will be assigned again (since a new variable is assigned for every function) in another function which will be different...

MaqOwais replied:
```suggestion
    integersomething = Symbol('integersomething',integer=True)
```
or else something ... I know it's hacky but some alternative for this should be there as the  similar functionality to symbol...

oscarbenjamin replied:
You just need to return the Dummy along with the expression. You shouldn't create symbols in different functions and expect them to be equal.

MaqOwais replied:
> You just need to return the Dummy along with the expression. You shouldn't create symbols in different functions and expect them to be equal.

yep, the output is Dummy even after this intermediate variable...

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +179,85 @@ def _lambert(eq, x):
                 continue
-            rhs = -c/b + (a/d)*w
+            sol1 = lambert_helper(rhs2)
+        return sol1,rhs2
```
Use space after comma

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -374,7 +457,12 @@ def _solve_even_degree_expr(expr, t, symbol):
         if mainexp:
             lhs = collect(lhs, mainexp)
             if lhs.is_Mul and rhs != 0:
-                soln = _lambert(expand_log(log(lhs) - log(rhs)), symbol)
+                if domain.is_subset(S.Reals):
+                    soln = _lambert_real(expand_log(log(lhs) - log(rhs)), symbol)
+                    if soln is S.EmptySet:
+                        return S.EmptySet
+                else:
+                    soln = _lambert(expand_log(log(lhs) - log(rhs)), symbol)
```
Since this logic is repeated below. Maybe it should be moved into `_lambert` or something...

MaqOwais commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -2310,7 +2206,162 @@ def test_solve_logarithm():
+    assert dumeq(solveset(exp(x) - 10*x, x), ImageSet(Lambda(n, -LambertW(Rational(-1, 10), n)), S.Integers))
+
+    assert solveset_real(exp(x) + x, x) == FiniteSet(-LambertW(1))
+    assert dumeq(solveset(exp(x) + x, x) , ImageSet(Lambda(n, -LambertW(1, n)), S.Integers))
```
you may see this one...
even after the symbol is used I make sure to return the dummy...
the point I want to make focus on is that if a `symbol` is used then it can use `subs` or `replace`  in between the functions whereas this can't be done if the dummy is used since we have to reassign the dummy variable in 
another function where it loses the identity and it regards in a different way(obviously this could be done within a function since we don't have to reassign dummy variable).....

MaqOwais committed:
removed integer symbol and refactored solveset

MaqOwais committed:
Update solveset.py

MaqOwais committed:
Update solveset.py

MaqOwais commented:
@oscarbenjamin please let me know if any further mistakes are present.
I think most of the things are covered up to this point.

MaqOwais commented:
@oscarbenjamin please let me know if any further mistakes are present.
I think most of the things are covered up to this point.

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +185,95 @@ def _lambert(eq, x):
+        lambert_real = LambertW(arg).is_real
+
+        if domain_real or domain is None:
+            if domain_real:
```
Rather than this:
```
if domain_real or domain is None:
    if domain_real:
        ...
    if domain is None:
```
you can just do:
```
if domain_real:
    ...
elif domain is None:
   ...
```

oscarbenjamin replied:
> I did this since there are common lines of code between None and Reals...

I don't see any common code here.

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +185,95 @@ def _lambert(eq, x):
+                        return S.EmptySet
+
+                if (arg.has(Symbol)) and not arg.has(Dummy):
+                    SymbolStatus = list(arg.atoms(Symbol))[0]
```
SymbolStatus is not a good variable name. Capitalised names are used for class names. A variable should have a lower-case name unless it is a mathematical symbol like `M`.

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +185,95 @@ def _lambert(eq, x):
+                        sol,rhs = lambert_braches()
+                else:
+                    rhs = rhs1
+            elif domain is None:
+                if not len(list(eq.atoms(Symbol,Dummy))) >1:
+                    if -1/E <= re(arg) < 0 and not arg.has(I):
+                        sol,rhs = lambert_braches()
+                    else:
+                        rhs = rhs1
+                else:
+                    rhs = rhs1
+            else:
+                rhs = rhs1
```
The chain of `rhs=rhs1` here suggests that the logic here can be rearranged so that it is cleaner.

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -151,7 +154,13 @@ def _lambert(eq, x):
+        xusolns = solve(X1 - u, x)
+    else:
+        domain_real = domain.is_subset(S.Reals)
+        xusolns = solvify(X1 - u, x,domain)
```
There should generally be a space after a comma (this applies elsewhere as well)

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -508,3 +597,4 @@ def ok(f, v, c):
             if new is not None:
                 return a*x*y + b*y, new, u
             x, y = y, x
+from sympy.solvers.solveset import solvify
```
Does this import need to be at the bottom because of circular imports? If so then that's fine. Otherwise it should go at the top of the file.

There should be two blank lines above the import line though.

oscarbenjamin commented on sympy/solvers/inequalities.py:
```diff
@@ -560,21 +543,40 @@ def valid(x):
+                    discontinuities = set(domain.boundary -
+                        FiniteSet(domain.inf, domain.sup))
+                    # remove points that are not between inf and sup of domain
+                    critical_points = FiniteSet(*(solns + singularities + list(
+                        discontinuities))).intersection(
+                        Interval(domain.inf, domain.sup,
+                        domain.inf not in domain, domain.sup not in domain))
```
This statement is horrible! Don't try to do so much in one statement. This can be several separate statements.

oscarbenjamin commented on sympy/solvers/inequalities.py:
```diff
@@ -560,21 +543,40 @@ def valid(x):
+                        domain.inf not in domain, domain.sup not in domain))
+                    if isinstance(critical_points,Intersection):
+                        raise ValueError
+            except (ValueError, NotImplementedError):
```
Using `try/except` should be contained to the smallest possible block of code. Ideally a `try` should only have one statement in it.

oscarbenjamin replied:
What is the reason for catching these exceptions anyway?

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -1753,6 +1777,182 @@ def _is_logarithmic(f, symbol):
+            if arg0.has(HyperbolicFunction,TrigonometricFunction,Dummy,Symbol)\
+                 and isinstance(arg0,(Dummy,Symbol,Pow)):
+                    j += 1
+        return j> 0
```
Maybe instead of `j += 1` there can just be `return True`

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -2208,16 +2437,18 @@ def solvify(f, symbol, domain):
     We classify the output based on the type of solution returned by `solveset`.
 
-    Solution    |    Output
+    Solution     |    Output
```
Why has this changed?

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -3191,7 +3428,11 @@ def _solve_using_known_values(result, solver):
                         # and add this new solution
                         if soln_imageset:
                             # replace all lambda variables with 0.
-                            imgst = soln_imageset[sol]
+                            try:
+                                imgst = soln_imageset[sol]
+                            except KeyError:
+                                continue
+
```
Why has this changed?

oscarbenjamin commented:
There are merge conflicts in this PR now. Those will need to be fixed before the tests can run again.

MaqOwais commented:
https://github.com/sympy/sympy/pull/20824#discussion_r579502589

I did this since there are common lines of code between `None` and `Reals`...

https://github.com/sympy/sympy/pull/20824#discussion_r579504145

Yes, there is a circular import ...

https://github.com/sympy/sympy/pull/20824#discussion_r579504691  , https://github.com/sympy/sympy/pull/20824#discussion_r579505264

The code statement was not written by me, I've just rearranged that part ( since prior to this commit the intersection from `solvify` was returning null so to maintain such logic I just rearranged that part )...


https://github.com/sympy/sympy/pull/20824#discussion_r579506524

since after modifying the `solvify` now, it returns solution for intersection set also...So to look vivid I've just added space.

oscarbenjamin commented:
It's much easier to read what's going on if you respond directly underneath each comment rather than writing a comment that links to the others.

MaqOwais commented:
> It's much easier to read what's going on if you respond directly underneath each comment rather than writing a comment that links to the others.

Actually, there is no option to do so

![Screen Shot 2021-02-20 at 8 45 33 PM](https://user-images.githubusercontent.com/57854867/108600326-9f625680-73bc-11eb-9dc1-4aee5fe26976.png)
...

MaqOwais commented:
https://github.com/sympy/sympy/pull/20824#discussion_r579506683

During the lambert analysis with `nonlinsolve` it gives `KeyError` {here the replacement of lambda variable with `0` so it give KeyError} ... 
e.g.
`ImageSet(Lambda(n, 2*LambertW(-y/2, n)), S.Integers)`

MaqOwais committed:
Merge branch 'master' into testing-other-stuff

MaqOwais committed:
refactored

MaqOwais committed:
Update solveset.py

MaqOwais committed:
Merge branch 'testing-other-stuff' of https://github.com/MaqOwais/sympy into testing-other-stuff

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +185,85 @@ def _lambert(eq, x):
-
     # calculating solutions from args
-    for arg in args:
+    if p == 1:
+        t = e
+        args = [d/(a*b)*t]
+    elif domain is S.Reals:
+        ind_ls = [d/(a*b)*t for t in roots(t**p - e, t).keys()]
+        args = []
+        j = -1
+        for i in ind_ls:
+            j += 1
+            if not isinstance(i,int):
+                if not i.has(I):
+                    args.append(ind_ls[j])
+            elif isinstance(i,int):
+                args.append(ind_ls[j])
+    else:
+        args = [d/(a*b)*t for t in roots(t**p - e, t).keys()]
```
There are many places in this PR where the code is unnecessarily convoluted. For example I think all of the code above can just be replaced with:
```python
root_list = list(roots(t**p - e, t))
if domain is S.Reals:
    root_list = [r for r in root_list if not r.has(I)]
args = [r*d/(a*b) for r in root_list]
```

MaqOwais committed:
refactored

MaqOwais commented:
> There are many places in this PR where the code is unnecessarily convoluted. For example I think all of the code above can just be replaced with:

Yes, I do believe that there are lot of unease in code written by me and also I regret this a lot and I'm trying my level best to complete this issue in a perfect way....N thank's a lot for your attention and guidance : )

MaqOwais commented:
@oscarbenjamin, I think in `sympy/solvers/bivariate.py`, `sympy/solvers/inequalities.py `, and `sympy/solvers/solvers.py `  most of the points are covered, and maybe in `sympy/solvers/solveset.py` there may be some convoluted code!
Kindly suggest me something to make this PR more appealing so that to proceed for merge.

MaqOwais committed:
better code ...

MaqOwais committed:
Update solveset.py

MaqOwais committed:
Update solveset.py

MaqOwais committed:
improved solveset.py and doc's

MaqOwais commented:
IMHO, I think most of the code is cleaned up ... 
Please, let me know if something is left up : )

Thanks...

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +185,73 @@ def _lambert(eq, x):
+        lambert_real = LambertW(arg).is_real
+
+        if domain_real:
+            if not len_of_symbol >1:
```
There should be a space after an operator like `>` but also it would make more sense to express this as
```
if len_of_symbol <= 1:
```
or whatever it is supposed to be.

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -2256,6 +2498,12 @@ def solvify(f, symbol, domain):
     elif isinstance(solution_set, FiniteSet):
         result = list(solution_set)
 
+    elif isinstance(solution_set, Intersection):
+        result = []
+        for argSet in solution_set.args:
+            if isinstance(argSet,FiniteSet):
+                result += list(argSet)
```
What does this do?

For example what would `solution_set` be and what would `result` end up being?

MaqOwais replied:
This converts Intersection set into a list ... (i.e. same quality as of `solve` function)

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -3196,7 +3444,12 @@ def _solve_using_known_values(result, solver):
                         # and add this new solution
                         if soln_imageset:
                             # replace all lambda variables with 0.
-                            imgst = soln_imageset[sol]
+                            try:
+                                imgst = soln_imageset[sol]
+                            # imageset(lambda(),) with 0 returns keyerror
+                            except KeyError:
+                                continue
```
I don't understand why this is here. Why wouldn't `sol` be in `soln_imageset`?

MaqOwais replied:
Actually here the evaluation of `imageset` is done if other than `Imageset` enter here then it shows error...
So, either `KeyError ` should be there or `if soln_imageset and sol.has(Dummy):` is to be used (at line 3303 and 3445).

MaqOwais replied:
> if other than Imageset enter here then it shows error

Since both `non-lambert` and `lambert` solutions are input to this logic, so both of them entering into this loop hence for non-lambert (where `imageset` is not present) should be diverted ....

MaqOwais replied:
> Please suggest any of the two here ...

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -3259,7 +3518,9 @@ def _solve_using_known_values(result, solver):
         # eg : [{x: -1, y : 1}, {x : -y , y: y}] then
         # return [{x : -y, y : y}]
         result_all_variables = result_infinite
-    if intersections or complements:
+    # not including lambert part since complex part has ImageSet for integers
+    # which is not compatable to find intersection
+    if (intersections or complements) and not has_lambert:
```
Why is this needed?

MaqOwais replied:
`if (intersections or complements) and not has_lambert:`
This is for getting the common intersection with different intervals with a different set of solutions... as lambert complex has Imageset and infinite solutions I think it's not compatible and also it's not returning the correct solution if `if (intersections or complements) and not has_lambert:` is not used..

MaqOwais replied:
I'll try to look more into this part ....

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -176,22 +185,73 @@ def _lambert(eq, x):
+        return S.EmptySet
+    # for simplicity defined two functions
+    # Lambert_Real_Braches , lambert_helper
+    def lambert_braches():
```
It should be `lambert_branches`. The comment above refers to `Lambert_Real_Braches` which I don't think is here any more.

MaqOwais replied:
oh yes...

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -1754,6 +1781,187 @@ def _is_logarithmic(f, symbol):
+    elif soln is None:
+     # try with positive `symbol`
+        u = Dummy('u', positive=True)
+        pos_lhs = lhs.subs({symbol: u})
```
It shouldn't be necessary to do this.

MaqOwais replied:
This is actually working and is underuse when the solution is None. Here trying to get the solution with a positive dummy!

MaqOwais replied:
Btw, this code is written in https://github.com/sympy/sympy/pull/14972 ...

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -1948,6 +2157,12 @@ def add_type(lhs, rhs, symbol, domain):
+            result = _solve_as_lambert(lhs, rhs, symbol, domain)
+            if isinstance(result, ConditionSet):
+                if _is_bivariate(lhs, symbol):
+                    result = _solve_as_bivariate(lhs, rhs, symbol, domain)
```
Why is `solve_as_bivariate` separate from `solve_as_lambert`?

MaqOwais replied:
I did not get your point, 
Actually according to the logic of code `solve_as_bivariate` is a subset of `_is_lambert` ( (`L-2160`) that's why it's under `elif _is_lambert(lhs, symbol):`   (`L-2160`)  ), so I think it is not separate but it's trying to work in different way ( see  `_solve_as_bivariate ` function) ....

MaqOwais replied:
Btw, this code is written in https://github.com/sympy/sympy/pull/14972 ...

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -2139,8 +2361,26 @@ def solveset(f, symbol=None, domain=S.Complexes):
         f = f.xreplace({d: e})
     f = piecewise_fold(f)
 
-    return _solveset(f, symbol, domain, _check=True)
+    result = _solveset(f, symbol, domain, _check=True)
+    if _is_lambert(f, symbol) and isinstance(result,FiniteSet):
+        # it could be extended to other than lambert
```
It shouldn't be necessary to make a special case for the lambert solver here in the main solveset function

MaqOwais commented:
Should move for the commit (hopefully a final one), with the above corrections and explanations?

oscarbenjamin commented:
I didn't realise this was based on #14972.

When I look through the diff there it seems much cleaner than this PR. Why have so many more things been changed compared to #14972?

MaqOwais commented:
> When I look through the diff there it seems much cleaner than this PR. Why have so many more things been changed compared to #14972?

Because that PR did not return
1. Imageset for lambert complex 
2. the branches (i.e. returned only lambertW(a,0) not lambertW(a,-1) always)
3. Domains were not specified 
4. returned both non-complex and complex ( containing `I`) in solution
5. didn't return `EmptySet` if lambert is complex i.e.   a < -1/E ( where lambertW(a,0) )
6. didn't return solution for such `solveset_real(x**2*exp(x**2) - x, x)`..
etc....

MaqOwais commented:
> When I look through the diff there it seems much cleaner than this PR. Why have so many more things been changed compared to #14972?

I've learned a lot of things in this process like 
1. planning, discussing a lot before execution till the end,
2. not just to write a hacky solution instead try to dive deeper for each logic around the code, 
3. not get frustrated early,
4. code quality, i.e., not using a global variable ( which was after tremendous work), unnecessary flags or not to disturb other functionality (i.e. solve ), try to avoid lengthy code, 
etc...

N I'm very grateful you've been with me on this journey ...

oscarbenjamin commented:
> Because that PR did not return

Okay, I see now.

The difficulty I have with this PR is the code is complicated and messy which makes it hard to tell if it is correct or not. For examples the `_is_lambert` function is very long and messy. I'm sure it can be written in a way that is simpler and possibly less buggy. The problem is I don't understand what the code there does to be able to suggest how to change it. The docstring says:
```
    Return True if the equation is of Lambert type, else False.
    Equations containing more than two variables involving
    `Pow`, `HyperbolicFunction`,`TrigonometricFunction`, `log` or `exp`
    terms are currently treated as Lambert types.
```
To me that suggests:
```
def is_lambert(eq):
    lambert_function_types = [Pow, log, exp, HyperbolicFunction, TrigonometricFunction]
    return len(eq.free_symbols) >= 2 and eq.has(lambert_function_types)
```
That's two lines but instead the function in this PR has over 60 lines. What is that function even doing? (The docstring should really explain this)

Also making changes in many parts of the codebase makes it much more difficult to understand what effect all those changes have. For example why are there changes in inequalities.py?

MaqOwais commented:
> why are there changes in inequalities.py?

I repeat there are no changes in the inequalities.py (instead of 2 - 3 lines of code) line but it's just rearranged...

oscarbenjamin commented:
> > why are there changes in inequalities.py?
> 
> I repeat there are no changes in the inequalities.py (instead of 2 - 3 lines of code) line but it's just rearranged...

Okay I found your previous comment:

> i. just rearranged the lines of code. (since at present solvify is returning solutions
> for intersection set whereas previously it was returning [] hence was raising the ValueError)

What exactly are the inputs and outputs to `solvify` in this case?

It shouldn't return an `Intersection`.

MaqOwais commented:
> > i. just rearranged the lines of code. (since at present solvify is returning solutions
> > for intersection set whereas previously it was returning [] hence was raising the ValueError) 

> What exactly are the inputs and outputs to solvify in this case?
> 
> It shouldn't return an Intersection.

since at present `solvify` is returning solutions for `Intersection` set whereas previously it was returning [] so it was raising the ValueError (in ` inequalities.py`), to maintain the same logic I modified the` inequalities.py`...

oscarbenjamin commented:
> since at present `solvify` is returning solutions for the `Intersection` set whereas previously it was returning [] so it was raising the ValueError (in ` inequalities.py` due to emptylist), to maintain the same logic I've modified the` inequalities.py`...

Can you give an example? What equation do you solve with `solvify` to get an `Intersection`?

MaqOwais commented:
> What equation do you solve with solvify to get an Intersection?

`solvify` doesn't give an intersection instead it returns a list of solutions for intersection set from `solveset`...(see `solvify` function...where `solveset` is giving solutions to `solvify` in form of set and then it's converted into `list` {in the same format as of `solve`})

MaqOwais committed:
lambert imp.

oscarbenjamin commented:
> > What equation do you solve with solvify to get an Intersection?
> 
> `solvify` doesn't give an intersection instead it returns a list of solutions for intersection set (from `solveset`)...(see `solvify` function...where `solveset` is giving solutions to `solvify` in form of set and then it's converted into `list` {in the same format as of `solve`})

Can you give an actual example explaining why the changes in inequalities.py are needed?

MaqOwais commented:
> > > What equation do you solve with solvify to get an Intersection?
> > 
> > 
> > `solvify` doesn't give an intersection instead it returns a list of solutions for intersection set (from `solveset`)...(see `solvify` function...where `solveset` is giving solutions to `solvify` in form of set and then it's converted into `list` {in the same format as of `solve`})
> 
> Can you give an actual example explaining why the changes in inequalities.py are needed?

Ok, I'll try to give example also please go through this.... https://github.com/sympy/sympy/pull/20824#issuecomment-786022115
....

MaqOwais commented:
before this was the case and to maintain the same,... the changes were made in `inequalities.py` (by maintaining same logic of returning [] from `solvify`)
```
zero = cos(1)**2 + sin(1)**2 - 1
raises(NotImplementedError, lambda: isolve(x**2 < zero, x))
raises(NotImplementedError, lambda: isolve(
    x**2 < zero*I, x))
```
if the changes were not made (after getting the solution from `solvify`) this would have been the output...
```
zero = cos(1)**2 + sin(1)**2 - 1
raises(TypeError, lambda: isolve(x**2 < zero, x))
raises(TypeError, lambda: isolve(
    x**2 < zero*I, x))
```

MaqOwais commented:
e.g.
this gives error 
`solveset((r**2)*((cos(t) - 4)**2)*sin(t)**2 < 9, r, S.Reals)`

MaqOwais commented:
2nd ex:

```
>>> from sympy import *
>>> from sympy.stats import Exponential,P
>>> from sympy.abc import x,y,z,a,b,c
>>> rate = Symbol('lambda', positive=True)
>>> p = Symbol("p", positive=True, real=True, finite=True)
>>> X = Exponential('x', rate)
>>> P(X > z)
#solns -- with old code getting error from solvify(due to intersection was returning [])--> None
exp(-lambda*z)
```
after that, it raises an error because it gives an error!( same as above)

MaqOwais commented on sympy/solvers/inequalities.py:
```diff
@@ -560,21 +543,40 @@ def valid(x):
+                    discontinuities = set(domain.boundary -
+                        FiniteSet(domain.inf, domain.sup))
+                    # remove points that are not between inf and sup of domain
+                    critical_points = FiniteSet(*(solns + singularities + list(
```
Actually, the `critical_points ` are converted to `Intersection` due to the `solns`(which are o/p of `solvify` see line L-557 ) which is the reason for `TypeError`,...

MaqOwais commented on sympy/solvers/inequalities.py:
```diff
@@ -560,21 +543,40 @@ def valid(x):
-                    discontinuities))).intersection(
-                    Interval(domain.inf, domain.sup,
-                    domain.inf not in domain, domain.sup not in domain))
                 if all(r.is_number for r in critical_points):
```
Further, here `critical_points ` is an `intersection` set that is ultimately responsible for TypeError

MaqOwais commented:
I hope I've answered the reason : )

oscarbenjamin commented:
What you're saying is that in some situations `critical_points` will be an `Intersection` and this will raise:
```
if all(r.is_number for r in critical_points):
```
I think that the code should not attempt to loop over `cirtical_points` if it is not a `FiniteSet`.

MaqOwais commented:
> What you're saying is that in some situations `critical_points` will be an `Intersection` and this will raise:
> 
> ```
> if all(r.is_number for r in critical_points):
> ```
> 
> I think that the code should not attempt to loop over `cirtical_points` if it is not a `FiniteSet`.

Yes that's why `solvify` was not returning solutions for `intersection` set ....So, to make sure the same logic (ie. I think that the code should not attempt to loop over cirtical_points if it is not a FiniteSet.) the rearranged/changes in `inequalities.py`.....

MaqOwais commented:
So IMHO,  the changes in `inequalities.py` should not be changed....

MaqOwais commented:
I think they're 2 ways to achieve that ....
1. The present way 
2. To make another function to address the same ( but that will be same code as of solivify upto some extent and some code extension (that may also lead to code redundancy ...)

MaqOwais commented:
> 2nd ex:
> 
> ```
> >>> from sympy import *
> >>> from sympy.stats import Exponential,P
> >>> from sympy.abc import x,y,z,a,b,c
> >>> rate = Symbol('lambda', positive=True)
> >>> p = Symbol("p", positive=True, real=True, finite=True)
> >>> X = Exponential('x', rate)
> >>> P(X > z)
> #solns -- with old code getting error from solvify(due to intersection was returning [])--> None
> exp(-lambda*z)
> ```
> 
> after that, it raises an error (if no change in code as of now is there) !( same as above)

And there are more 7 examples that are presently failing if the code in `inequalities.py` remains the same (i.e. If code not rearranged) ....

oscarbenjamin commented:
It's better to check for a FiniteSet than catch TypeError. There are far too many places in sympy that catch TypeError. A TypeError should be considered a bug so we should not try to catch it: it is better to fix the bug so that TypeError is not raised.

MaqOwais commented on sympy/solvers/inequalities.py:
```diff
@@ -560,21 +543,40 @@ def valid(x):
+                        Interval(domain.inf, domain.sup,
+                        domain.inf not in domain, domain.sup not in domain))
+                    if isinstance(critical_points,Intersection):
+                        raise ValueError
```
```suggestion
                        solns = []
```

MaqOwais replied:
```
                solns = solvify(e, gen, domain)

                if solns:
                    discontinuities = set(domain.boundary -
                        FiniteSet(domain.inf, domain.sup))
                    # remove points that are not between inf and sup of domain
                    critical_points = FiniteSet(*(solns + singularities + list(
                        discontinuities))).intersection(
                        Interval(domain.inf, domain.sup,
                        domain.inf not in domain, domain.sup not in domain))
                    if isinstance(critical_points,Intersection):
                        solns = []
                if solns is None:
                    # in which case we raise ValueError
                    raise ValueError
```
Is this the correct way to proceed ?

MaqOwais replied:
The remaining code about raising errors was not written by me...
only the below line of code are addition (+30 -28) remaining are just rearrangements...

```
if isinstance(critical_points,Intersection):
           solns = []
```

oscarbenjamin replied:
I think it would be better to test `if isinstance(critical_points, FiniteSet)`. That way we test for a case that we know we can handle. Also setting `solns = []` is cryptic. It would be better to test closer to where the test is needed.

MaqOwais replied:
> I think it would be better to test `if isinstance(critical_points, FiniteSet)`. That way we test for a case that we know we can handle. Also setting `solns = []` is cryptic. It would be better to test closer to where the test is needed.

At this point, we're mostly concerned about not to get `intersection` set (https://github.com/sympy/sympy/pull/20824#issuecomment-786835550)

> Also setting `solns = []` is cryptic

So, what about `solns = None`( if it's intersection then `solns` should be None). Also, the `solns` is not used anywhere in the function except in `critical_points`.

MaqOwais replied:
@oscarbenjamin 
please give your view on this (https://github.com/sympy/sympy/pull/20824#discussion_r584020497)

MaqOwais replied:
> > I think it would be better to test `if isinstance(critical_points, FiniteSet)`. That way we test for a case that we know we can handle. Also setting `solns = []` is cryptic. It would be better to test closer to where the test is needed.
> 
> At this point, we're mostly concerned about not to get `intersection` set ([#20824 (comment)](https://github.com/sympy/sympy/pull/20824#issuecomment-786835550))
> 
> > Also setting `solns = []` is cryptic
> 
> So, what about `solns = None`( if it's intersection then `solns` should be None). Also, the `solns` is not used anywhere in the function except in `critical_points`.
> 
> I think the `solns` variable should be renamed to something like `critical_solns` so that it will be more specific ( where using `solns` give a broad sense. )

Or else ,... what about taking

```
if not isinstance(critical_points,FiniteSet):
    solns = None
```

oscarbenjamin replied:
> Or else ,... what about taking
> 
> ```
> if not isinstance(critical_points,FiniteSet):
>     solns = None
> ```

I think that would probably be better.

MaqOwais replied:
> I think the solns variable should be renamed to something like critical_solns so that it will be more specific ( where using solns give a broad sense. )

And, what about this?

MaqOwais reviewed (commented):
Kindly, give your opinion on this ....

MaqOwais committed:
improved inequalities

MaqOwais commented:
Is there any update on this?

oscarbenjamin commented:
When I look at a PR what I look at is the diff here:
https://github.com/sympy/sympy/pull/20824/files

I want to go through that and make sure I understand each of the changes. The fewer things you change the better. I still find looking at inequalities.py that I don't understand why so many things have changed when this PR shouldn't have anything to do with inequalities. That makes me suspicious that something is not working correctly but I can't tell what.

Can you try to change as few things as possible and keep the diff as small as possible?

If it's possible to just change one line in inequalities.py then it will be much easier for me to understand what the change does.

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -1754,6 +1781,148 @@ def _is_logarithmic(f, symbol):
+    `Pow`, `HyperbolicFunction`,`TrigonometricFunction`, `log` or `exp`
+    terms are currently treated as Lambert types.
+    """
+    term_factors = list(_term_factors(f.expand()))
+
+    lambert_condition1 = len(Add.make_args(f)) >= 2 and len([arg for arg in term_factors if arg.has(Symbol)]) >= 2
+    lambert_condition2 = any(isinstance(arg, (log, HyperbolicFunction, TrigonometricFunction)) and arg.has(Symbol) for arg in term_factors)
+    lambert_condition3 = any(isinstance(arg, (Pow, exp)) and arg.has(Symbol) for arg in term_factors if (arg.as_base_exp()[1]).has(Symbol))
+
+    return lambert_condition1 and lambert_condition2 or lambert_condition3
```
These lines are too long. It would be clearer if you wrote this as:
```
if len(...) > 2:
    if an(...):
        return True
    if any(...):
        return True
return False
```
The variable name `lambert_condition1` tells us nothing about what it means. Should be `has_two_terms` or something?

Why are these particular conditions being checked? Why check `Add.make_args` and `term_factors` as well?

Do you think that someone looking at this code would understand what it is doing?

The code should be written more clearly and if it is still not obvious what it is doing then there should be comments explaining what it does. Also the docstring should match what the function actually does.

MaqOwais replied:
> Should be has_two_terms or something?
> Why are these particular conditions being checked? Why check Add.make_args and term_factors as well?

yes but not just only 2 variables...also to have at least 2 operands and also two variables in each term_factor...

MaqOwais replied:
> Should be has_two_terms or something?
> Why are these particular conditions being checked? Why check Add.make_args and term_factors as well?

yes but not just only 2 variables...also to have at least 2 operands and also two variables in each term_factor...

MaqOwais replied:
> Do you think that someone looking at this code would understand what it is doing?

Oh! sorry for this, I'll try better next time!

MaqOwais replied:
And the reason for using `lambert_condition3` and `lambert_condition2` 
In, `exp and pow `terms the exponent should have a symbol whereas this is not valid for  `log, HyperbolicFunction, TrigonometricFunction`

MaqOwais replied:
And also `len(eq.free_symbols)` can't be used because 

```
>>> _is_lambert(3*log(x) - x*log(3))
{x} f.free_symbols
False
```

MaqOwais replied:
```suggestion
    term_factors = list(_term_factors(f.expand()))
    # total number of symbols in equation
    no_of_symbols = len([arg for arg in term_factors if arg.has(Symbol)])
    no_of_operands = len(Add.make_args(f))

    if no_of_operands >= 2 and no_of_symbols >= 2 :
        # `log, HyperbolicFunction, TrigonometricFunction` should have symbols
        if any(isinstance(arg, (log, HyperbolicFunction, TrigonometricFunction))\
            for arg in term_factors if arg.has(Symbol)):
            return True
        # here, `Pow, exp` exponent should have symbols
        elif any(isinstance(arg, (Pow, exp)) \
            for arg in term_factors if (arg.as_base_exp()[1]).has(Symbol)):
            return True
    return False
```

MaqOwais commented on sympy/solvers/bivariate.py:
```diff
@@ -151,7 +154,13 @@ def _lambert(eq, x):
+        xusolns = solve(X1 - u, x)
+    else:
+        domain_real = domain.is_subset(S.Reals)
+        xusolns = solvify(X1 - u, x, domain)
```
>  I still find looking at inequalities.py that I don't understand why so many things have changed when this PR shouldn't have anything to do with inequalities. 

Please try to understand that `solvify` is being used and is common for both and the `solvify` functions is also modified in the way so that it should give the solutions to `intersection` set also....

MaqOwais commented on sympy/solvers/solveset.py:
```diff
@@ -2256,6 +2441,12 @@ def solvify(f, symbol, domain):
     elif isinstance(solution_set, FiniteSet):
         result = list(solution_set)
 
+    elif isinstance(solution_set, Intersection):
+        result = []
+        for argSet in solution_set.args:
+            if isinstance(argSet,FiniteSet):
+                result += list(argSet)
```
here are the modifications in order to give the solution for the `intersection` set...

MaqOwais committed:
removed changes in inequalities

MaqOwais commented:
I think this commit is a good one ...

MaqOwais commented:
I think this commit is a good one ... Can you please add something to this...
There is another way also for this but that will be a little more recursive, but I feel that this commit has redundant code...

oscarbenjamin commented:
Yes, that's better.

oscarbenjamin commented on sympy/solvers/bivariate.py:
```diff
@@ -151,7 +157,17 @@ def _lambert(eq, x):
 
     # invert the generator X1 so we have x(u)
     u = Dummy('rhs')
-    xusolns = solve(X1 - u, x)
+    if domain is None:
+        # the flow from solve will proceed from here
+        domain_real = None
+        xusolns = solve(X1 - u, x)
+    else:
+        # converting solutions from `solveset` into `list` format
+        # by having `periodicity` and passing it to `convert_to_list`
+        xusolns_1 = solveset(X1 - u, x, domain)
+        period = periodicity(X1 - u, x)
+        xusolns = convert_to_list(xusolns_1, period, domain)
+        domain_real = domain.is_subset(S.Reals)
```
Why does this use solve in one case and solveset in the other?

MaqOwais commented:
> Yes, that's better.

The one which is committed...? or the other possible way in review ?

oscarbenjamin commented:
> > Yes, that's better.
> 
> The one which is committed...? or the other possible way in review ?

The fact that there are no changes to inequalities.py now in this PR

MaqOwais commented:
> because we can't pass domain to solve so that's why it's necessary...(and also # the flow from solve will proceed from here (for domain is None) since sols = _solve_lambert(_eq, sym, g, None) (L -2704) in solve.py module)

I hope this is clear...

MaqOwais commented:
Can you please give any further reviews, so that I could speed up the process and finish this PR? 
And further, spend more time on other PRs.
Thank you : )

oscarbenjamin commented:
> Can you please give any further reviews, so that I could speed up the process and finish this PR?

I don't think you understand how time consuming it would be for me to try and understand all of this code. I'm sure that the changes here are more complicated than they need to be but telling you exactly how to change it would be  as much work as writing the code in the first place.

What would make this faster would be if you:

1. Break changes into smaller PRs.
2. Change as few things as possible (keep the diff small).
3. Try to simplify your own code as much as possible before asking anyone else to review it.
4. Carefully explain what each change is doing and why both in comments and in the PR.

Look for example at your description in the OP of this PR:
```
Tried to make a better version according to the latest updates
as well as removing some errors accordingly
```
How is anyone supposed to understand what that means? Should they go and read all of the discussion in the 6 linked issues to understand what the purpose of this PR is?

MaqOwais commented:
> How is anyone supposed to understand what that means? Should they go and read all of the discussion in the 6 linked issues to understand what the purpose of this PR is?

Oh, I understand ! I've to also put a lot of focus on such things also ( to deliver the message with a lot of ease)...
